server/server.go

```diff
@@ -254,3 +259,20 @@ func mustRegisterGWHandler(register registerFunc, ctx context.Context, mux *runt
 		panic(err)
 	}
 }
+
+// Authenticate checks for the presence of a token when accessing server-side resources.
+func (a *ArgoCDServer) authenticate(ctx context.Context) (context.Context, error) {
+	if md, ok := metadata.FromIncomingContext(ctx); ok {
+		mgr := util_session.MakeSessionManager(a.settings.ServerSignature)
+		tokens := md["tokens"]
+		for _, token := range tokens {
+			_, err := mgr.Parse(token)
+			if err == nil {
+				return ctx, nil
+			}
+		}
+		return ctx, fmt.Errorf("user is not allowed access")
+	}
+
+	return ctx, fmt.Errorf("empty metadata")
```

Comment:
This interceptor should return grpc errors instead of generic go errors. Specifically it should be:

grpc.Errorf(codes.Unauthenticated, message)